Custom-Audio-Driver/Custom-Audio-Driver/OTDefaultAudioDevice-Mac.m

```diff
@@ -711,4 +736,202 @@ - (BOOL)setPlayOutRenderCallback:(AudioUnit)unit
+                                               0,
+                                               &deviceID,
+                                               sizeof(deviceID));
+        NSError *error = [NSError errorWithDomain:NSOSStatusErrorDomain code:status userInfo:nil];
+        NSLog(@"error: %@", error.localizedDescription);
+    }
+}
+
++ (AudioDeviceID)getDefaultAudioOutputDeviceID
+{
+    // Get the default output device.
+    AudioDeviceID deviceID;
+    UInt32 defaultOutputPropSize = sizeof(AudioDeviceID);
+    AudioObjectPropertyAddress defaultOutputAddress = {
+        kAudioHardwarePropertyDefaultOutputDevice,
+        kAudioObjectPropertyScopeGlobal,
+        kAudioObjectPropertyElementMain,
+    };
+    OSStatus status = AudioObjectGetPropertyData(kAudioObjectSystemObject,
+                                                 &defaultOutputAddress,
+                                                 0,
+                                                 NULL,
+                                                 &defaultOutputPropSize,
+                                                 &deviceID);
+    if (status != noErr) {
+        NSError *error = [NSError errorWithDomain:NSOSStatusErrorDomain code:status userInfo:nil];
+        NSLog(@"error: %@", error.localizedDescription);
+    }
+
+    NSLog(@" - UID:             %@", getStringProperty(deviceID, kAudioDevicePropertyDeviceUID));
+    NSLog(@" - Model UID:       %@", getStringProperty(deviceID, kAudioDevicePropertyModelUID));
+    NSLog(@" - Name:            %@", getStringProperty(deviceID, kAudioDevicePropertyDeviceNameCFString));
+    NSLog(@" - Manufacturer:    %@", getStringProperty(deviceID, kAudioDevicePropertyDeviceManufacturerCFString));
+    NSLog(@" - Input channels:  %@", @(getChannelCount(deviceID, kAudioObjectPropertyScopeInput)));
+    NSLog(@" - Output channels: %@", @(getChannelCount(deviceID, kAudioObjectPropertyScopeOutput)));
+    NSLog(@" - Input source:    %@", getSourceName(deviceID, kAudioObjectPropertyScopeInput));
+    NSLog(@" - Output source:   %@", getSourceName(deviceID, kAudioObjectPropertyScopeOutput));
+    NSLog(@" - Transport type:  %@", getCodeProperty(deviceID, kAudioDevicePropertyTransportType));
+    NSLog(@" - Icon:            %@", getURLProperty(deviceID, kAudioDevicePropertyIcon));
+    return deviceID;
+}
+
+static inline AudioObjectPropertyAddress makeGlobalPropertyAddress(AudioObjectPropertySelector selector)
+{
+    AudioObjectPropertyAddress address = {
+        selector,
+        kAudioObjectPropertyScopeGlobal,
+        kAudioObjectPropertyElementMain,
+
+    };
+    return address;
+}
+
+static NSString *formatStatusError(OSStatus status)
+{
+    if (status == noErr) {
+        return [NSString stringWithFormat:@"No error (%d)", status];
+    }
+
+    return [NSString stringWithFormat:@"Error \"%s\" (%d)",
+            codeToString(status),
+            status];
+}
+
+static char *codeToString(UInt32 code)
```

Comment:
Out of curiosity, why is this needed? It seems to be only used to copy bytes from the uint32 to the string. Then it is always used together with the uint32.

Reply:
It's just for logging purposes.